### PR TITLE
Support send request in protocol server

### PR DESCRIPTION
**Pull request by testforstephen**
Signed-off-by: Jinbo Wang <jinbwan@microsoft.com>

### CODE REVIEW

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/IProtocolServer.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+
+package com.microsoft.java.debug.core.protocol;
+
+import java.util.function.Consumer;
+
+import com.microsoft.java.debug.core.protocol.Events.DebugEvent;
+import com.microsoft.java.debug.core.protocol.Messages.Request;
+import com.microsoft.java.debug.core.protocol.Messages.Response;
+
+public interface IProtocolServer {
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/IProtocolServer.java]: Perfect! Thanks.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java [hunk 2]
```diff
@@ -102,14 +117,80 @@ protected void sendMessage(Messages.ProtocolMessage message) {
...
+
+    @Override
+    public void sendRequest(Messages.Request request, Consumer<Messages.Response> cb) {
+        sendRequest(request, 0, cb);
+    }
+
+    @Override
+    public void sendRequest(Messages.Request request, long timeout, Consumer<Messages.Response> cb) {
+        Timer timer = new Timer();
+        Disposable[] disposable = new Disposable[1];
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java]: Is this is a new trick? Not quite understand this one...

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java [hunk 3]
```diff
@@ -102,14 +117,80 @@ protected void sendMessage(Messages.ProtocolMessage message) {
...
+        boolean[] executed = new boolean[]{false};
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java]: And this one...

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/IProtocolServer.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+     * Send a request to the DA. And create a timeout error response to the callback if no response is received at the give time.
+     *
+     * @param request
+     *            the request message.
+     * @param timeout
+     *            the maximum time (in millis) to wait.
+     * @param cb
+     *            the request call back function.
+     */
+    void sendRequest(Request request, long timeout, Consumer<Response> cb);
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/IProtocolServer.java]: One suggestion here. Can we use CompletableFuture<Response> as return value? This way we have the chance to also pass errors like a timeout exception.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/IProtocolServer.java [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+public interface IProtocolServer {
+    /**
+     * Send a request to the DA.
+     *
+     * @param request
+     *            the request message.
+     * @param cb
+     *            the request call back function.
+     */
+    void sendRequest(Request request, Consumer<Response> cb);
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/IProtocolServer.java]: One suggestion here. Can we use CompletableFuture as return value? This way we have the chance to also pass errors like a timeout exception.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java [hunk 6]
```diff
@@ -45,6 +53,8 @@
     private int contentLength = -1;
     private AtomicInteger sequenceNumber = new AtomicInteger(1);
 
+    private PublishSubject<Messages.Response> subject = PublishSubject.<Messages.Response>create();
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java]: subject -> responseSubject

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java [hunk 7]
```diff
@@ -102,14 +117,80 @@ protected void sendMessage(Messages.ProtocolMessage message) {
...
+    public void sendRequest(Messages.Request request, Consumer<Messages.Response> cb) {
+        sendRequest(request, 0, cb);
+    }
+
+    @Override
+    public void sendRequest(Messages.Request request, long timeout, Consumer<Messages.Response> cb) {
+        Timer timer = new Timer();
+        Disposable[] disposable = new Disposable[1];
+        boolean[] executed = new boolean[]{false};
+        disposable[0] = subject.filter(response -> response.request_seq == request.seq).subscribe((response) -> {
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java]: If we only need to get one response from the subject, we can use take(1) and we don't need to dispose the subscription that way.
subject.filter(...).take(1).subscribe(...)

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java [hunk 8]
```diff
@@ -102,14 +117,80 @@ protected void sendMessage(Messages.ProtocolMessage message) {
...
+            synchronized (executed) {
+                if (executed[0]) {
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java]: By using take(1), this won't be called twice.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java [hunk 9]
```diff
@@ -102,14 +117,80 @@ protected void sendMessage(Messages.ProtocolMessage message) {
...
+                            if (executed[0]) {
+                                return;
+                            }
+                            executed[0] = true;
+                        }
+                        if (disposable[0] != null) {
+                            disposable[0].dispose();
+                            disposable[0] = null;
+                        }
+                        cb.accept(new Messages.Response(request.seq, request.command, false, "timeout"));
```
Comment (akaroml) [inline com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java]: Reporting timeout error using the normal execution flow is not recommended. Please switch to CompletableFuture and try to completeExceptionally with a TimeoutException.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java (hunk 2 above)
Comment (testforstephen) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/AbstractProtocolServer.java]: when use Disposable directly in subscriber, the compile told me the local variable Disposable may not be initialized. So use array to workaround the issue.

File: com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/IProtocolServer.java (hunk 1 above)
Comment (markaw67) [reply com.microsoft.java.debug.core/src/main/java/com/microsoft/java/debug/core/protocol/IProtocolServer.java]: No problem

**Pull request merged**